:memo: Aiven integration (#2085)
FusionAuth/fusionauth-site: site/docs/v1/tech/samlv2/aiven.adoc

---
layout: doc
title: SAML v2 & Aiven
description: Setting up Aiven to login using FusionAuth via SAML v2
navcategory: login-methods
---

== Overview

Aiven is a cloud-based Database-as-a-Service platform. This document covers the configuration necessary to get it working with FusionAuth as the Identity Provider using SAML v2, allowing users to log in to your Aiven account using their FusionAuth credentials.

=== Prerequisites

This document assumes you have a running instance of FusionAuth and a working Aiven application with a project already created. You will also need admin accounts for both to configure them correctly.

[WARNING]
====

> This does not work well if you are running FusionAuth on Docker. Please check out https://github.com/FusionAuth/fusionauth-issues/issues/2308[this issue] for progress on fixing it.

Reviewer: ```suggestion
This does not work if you are running FusionAuth on Docker. Please check out https://github.com/FusionAuth/fusionauth-issues/issues/2308[this issue] for progress on fixing this problem.
```

====

== Configure Aiven

=== Create Team

First, you need to create a team where new users will be added to. In your https://console.aiven.io/[Aiven console], browse to [breadcrumb]#Admin -> Organization# and click [uielement]#Create team# in the organization details section. Set the [field]#Team name# as `Developers` and click [uielement]#Create team#.

When arriving back to the teams page, click on the newly created team and go to the [uielement]#Projects and Roles# tab. Click [uielement]#Add projects#, select the project, the desired [field]#Permission level# (e.g. `Developer`) and click [uielement]#Add project# to finish setting up the team.

image::samlv2/aiven/add-team-to-project.png[Adding team to a project.,width=1200]

=== Configure SSO

Click on [breadcrumb]#Authentication# in the sidebar and then on [uielement]#Add authentication method# to create one. Name it `FusionAuth`, select `SAML` as the [field]#Method Type# and choose the `Developers` team you just created in [field]#Team to autojoin on signup#. Click [uielement]#Add method# to save.

image::samlv2/aiven/add-authentication-method.png[Adding an authentication method.,width=1200]

> In the next screen, write down both [field]#Metadata URL# and [field]#ACS URL# values and click [uielement]#Finish#.

Reviewer: ```suggestion
In the next screen, record both the [field]#Metadata URL# and [field]#ACS URL# values and click [uielement]#Finish#. You'll use these values later.
```

== Configure FusionAuth

> === Generating Certificate

Reviewer: ```suggestion
=== Generating the Certificate
```

> You need to generate a custom RSA certificate and upload it to your FusionAuth instance's Key Master. To do so, there's a script in our https://github.com/FusionAuth/fusionauth-example-scripts[Example scripts GitHub repository] that can do all of this for you.

Reviewer: ```suggestion
You need to generate a custom RSA certificate and upload it to your FusionAuth instance's Key Master. To do so you can use a script in our https://github.com/FusionAuth/fusionauth-example-scripts[example scripts GitHub repository].
```

Before running it, you need to create an API Key in your FusionAuth instance. Browse to [breadcrumb]#Settings -> API Keys# and click on icon:plus[role=ui-button green,type=fas]. Set the [field]#Description# as `Certificate generator` and locate `/api/key/import` in the [uielement]#Endpoints# section. Click the `POST` switch to enable it and confirm by clicking on icon:save[role=ui-button blue,type=fas].

image::samlv2/aiven/create-api-key.png[Creating API Key.,width=1200]

> After being redirected back to the [breadcrumb]#API Keys# page, click on the `Key` for the created item to reveal its actual value and copy it. You'll have to provide this value to the script soon.

Reviewer: ```suggestion
After being redirected back to the [breadcrumb]#API Keys# page, click on the `Key` field to reveal its actual value. Then copy it. You'll have to provide this value to the script below.
```

image::samlv2/aiven/grab-api-key.png[Grabbing API Key.,width=1200,role=bottom-cropped]

> Now that your application is ready, clone our https://github.com/FusionAuth/fusionauth-example-scripts[Example scripts GitHub repository] and execute the script located in `v3-certificate`.

Reviewer: You could also remote include the file and have them copy paste it.

[source,shell]
----
git clone git@github.com:FusionAuth/fusionauth-example-scripts.git
cd fusionauth-example-scripts/v3-certificate
./generate-certificate
----

Answer the questions the script will ask you and make sure to give the key a meaningful name, like `Aiven key`.

[source,shell]
.Example response for the script
----
Your FusionAuth instance URL (with scheme): http://localhost:9011
API Key with /api/key/import endpoint:      Dy9bphElA3L3_ayW86T5KvrZkyK1Gj5EDV_2NKU5Ypg
Name for the generated key:                 Aiven key
----

> After finishing, you'll have a certificate in the Key Master in your FusionAuth instance ready to be used. The script will also print the generated certificate, so make sure to write it down.

Reviewer: ```suggestion
After finishing, you'll have a certificate in the Key Master in your FusionAuth instance ready to be used. The script will also print the generated certificate. Copy that as well, as you'll need it later.
```

> === Creating Application

Reviewer: ```suggestion
=== Creating the Application
```

Create an application in your FusionAuth instance by navigating to [breadcrumb]#Applications# and clicking on icon:plus[role=ui-button green,type=fas]. Name it `Aiven` and go to the [breadcrumb]#SAML# tab. Toggle the [field]#Enabled# switch, paste the [field]#Metadata URL# and [field]#ACS URL# you copied from Aiven to [field]#Issuer# and [field]#Authorized redirect URLs# fields in your FusionAuth application, respectively.

.Mapping fields
[cols="1,1"]
|===
|Aiven field|FusionAuth field

|[field]#Metadata URL#
|[field]#Issuer#

|[field]#ACS URL#
|[field]#Authorized redirect URLs#
|===

> Scroll down to the [uielement]#Authentication response# section and change the [field]#Signing key# to the `Aiven key` you created above. Click on icon:save[role=ui-button blue,type=fas] to save your application.

Reviewer: ```suggestion
Scroll down to the [uielement]#Authentication response# section and change the [field]#Signing key# to the `Aiven key` you imported above. Click on icon:save[role=ui-button blue,type=fas] to save your application.
```

> When redirected to the [breadcrumb]#Applications# page, view your application details by clicking on icon:search[role=ui-button green,type=fas]. In [uielement]#SAML v2 Integration details#, write down both [field]#Entity Id# and [field]#Login URL# fields.

Reviewer: ```suggestion
When redirected to the [breadcrumb]#Applications# page, view your application details by clicking on icon:search[role=ui-button green,type=fas]. In [uielement]#SAML v2 Integration details#, record both the [field]#Entity Id# and [field]#Login URL# fields.
```

== Finish Aiven Configuration

Go back to your https://console.aiven.io/[Aiven console] and click [breadcrumb]#Admin# and then on [breadcrumb]#Authentication# in the sidebar to list the authentication methods again. Click on `FusionAuth`, which will still have a `PENDING CONFIGURATION` state until you finish setting it up.

Scroll down to the [uielement]#SAML config# section and click the [uielement]#Edit# button and fill in the information below.

* [field]#SAML IDP Url#: the [field]#Login URL# from your FusionAuth application.
* [field]#SAML Entity ID#: the [field]#Entity Id# from your FusionAuth application.
* [field]#SAML Signature Algorithm#: you can leave this blank.
* [field]#SAML Digest Algorithm#: you can leave this blank.
* [field]#SAML Field Mapping#: you can leave this blank.

> * [field]#SAML Certificate#: paste the certificate you've copied from the `Generating certificate` two steps above.

Reviewer: ```suggestion
* [field]#SAML Certificate#: paste the certificate you've copied from the `Generating certificate` step above.
```

Finally, click [uielement]#Edit method# to save the changes.

image::samlv2/aiven/saml-configuration.png[Configuring SAML in Aiven.,width=1200]

> Go to the beginning of the page and click the [field]#Enable authentication method# switch to enable it. Now scroll down to [uielement]#Signup and link accounts URLs# and  copy the [uielement]#Signup URL#. You'll use it to test the integration and then share it with others to log into your Aiven application.

Reviewer: ```suggestion
Go to the beginning of the page and click the [field]#Enable authentication method# switch to enable logging in with FusionAuth.

Scroll down to [uielement]#Signup and link accounts URLs# and  copy the [uielement]#Signup URL#. You'll use it to test the integration and then share it with others to log into your Aiven application.
```

image::samlv2/aiven/copy-signup-url.png[Copy Signup URL from Aiven.,width=1200]

== Testing

Open the [uielement]#Signup URL# you copied above in an incognito tab or using another browser. You'll reach the signup page below.

image::samlv2/aiven/login-sso.png[Logging in to Aiven.,width=542,role=custom]

> Click on [uielement]#Sign up with FusionAuth#, which will redirect you to the FusionAuth login page. Fill in your credentials and submit the form to be taken back to the Aiven console, already logged in and part of the `Developers` team.

Reviewer: Do you need to register the user for the Aiven application?

Do you need to have any specific roles in FusionAuth that map to roles in Aiven?

Reviewer: @vcampitelli did you miss this?

Author: No, it was late and I decided to leave it for today, that's why I didn't re-request your review yesterday :smile: I just did it. 

And no, no specific roles are needed.
